File: 2019/23/solution.ts
import * as fs from "fs";
import * as path from "path";
import intcodeGenerator, { IntcodeGenerator } from "../intcode/intcode";

const puzzleInput = fs
  .readFileSync(path.join(__dirname, "input.txt"), "utf-8")
  .split(",")
  .map(Number);

const computers: Map<number, IntcodeGenerator> = new Map();

let outbox: number[][] = [];

// initialize computers with network addresses
for (let i = 0; i < 50; i++) {
  const computer = intcodeGenerator(puzzleInput);
  computer.next();
  computer.next(i);
  computers.set(i, computer);
}

function fillOutbox(
  outbox: number[][],
  computer: IntcodeGenerator,
  lastRet: IteratorResult<number, number>
) {
  // fills outbox with output from computer until output is -9999
  // (computer is done sending and is now requesting input)
  while (lastRet.value != -9999) {
    const dst = lastRet.value;
    const x = computer.next().value;
    const y = computer.next().value;
    outbox.push([dst, x, y]);
    lastRet = computer.next();
  }
}

let i = 0;
let ret: IteratorResult<number, number>;
let natPacket = [0, 0];
let part1 = 0;
const ySet = new Set();
while (true) {
  // iterate through computers and fill outbox
  for (let j = 0; j < 50; j++) {
    const computer = computers.get(j)!;
    ret = computer.next(-1);
    fillOutbox(outbox, computer, ret);
  }

  // if nothing in outbox, network is idle.
  // NAT sends its last received packet to computer 0.
  if (outbox.length === 0) {
    if (ySet.has(natPacket[1])) {
      // y value has been seen before, puzzle solved!
      console.log(part1);
      console.log(natPacket[1]);
      break;
    }
    ySet.add(natPacket[1]);
    outbox = [[0, ...natPacket]];
  }

  // iterate through outbox (outbox may grow during this process)
  while (i < outbox.length) {
    const [dst, x, y] = outbox[i];
    i++;
    if (dst === 255) {
      // only keep track of last packet sent to 255
      natPacket = [x, y];
      // only keep track of first y value sent to 255
      if (!part1) part1 = y;
      continue;
    }
    const computer = computers.get(dst)!;
    computer.next(x);
    ret = computer.next(y);
    fillOutbox(outbox, computer, ret);
  }

  // reset outbox
  outbox = [];
  i = 0;
}

File: 2019/23/solution.go
package main

import (
	"fmt"
	"os"
	"path/filepath"
	"runtime"
	"strconv"
	"strings"

	"github.com/weitecklee/adventofcode/2019/intcode"
)

func main() {
	_, filename, _, _ := runtime.Caller(0)
	dirname := filepath.Dir(filename)
	inputFilePath := filepath.Join(dirname, "input.txt")
	data, err := os.ReadFile(inputFilePath)
	if err != nil {
		panic(err)
	}
	puzzleInput := parseInput(strings.Split(string(data), ","))
	fmt.Println(part1(puzzleInput))
	fmt.Println(part2(puzzleInput))
}

func parseInput(data []string) []int {
	numbers := make([]int, 0, len(data))
	for _, s := range data {
		if n, err := strconv.Atoi(s); err != nil {
			panic(err)
		} else {
			numbers = append(numbers, n)
		}
	}
	return numbers
}

type Packet struct {
	dst int
	x   int
	y   int
}

func collectPackets(packets *[]Packet, ch chan int) {
	var ret, x, y int
	for {
		ret = <-ch
		if ret == intcode.REQUESTSIGNAL {
			break
		}
		x = <-ch
		y = <-ch
		*packets = append(*packets, Packet{ret, x, y})
	}
}

func part1(puzzleInput []int) int {
	network := make(map[int]chan int, 50)
	for i := range 50 {
		ch := make(chan int)
		ic := intcode.NewIntcodeProgram(puzzleInput, ch)
		go ic.Run()
		<-ch
		ch <- i
		network[i] = ch
	}
	var packet Packet
	var packets []Packet
	var ch chan int

	for i := range 50 {
		ch = network[i]
		<-ch
		ch <- -1
		collectPackets(&packets, ch)
	}

	for len(packets) > 0 {
		packet = packets[0]
		packets = packets[1:]
		if packet.dst == 255 {
			return packet.y
		}
		ch = network[packet.dst]
		ch <- packet.x
		<-ch
		ch <- packet.y
		collectPackets(&packets, ch)
	}

	return -1

}

func part2(puzzleInput []int) int {
	network := make(map[int]chan int, 50)
	for i := range 50 {
		ch := make(chan int)
		ic := intcode.NewIntcodeProgram(puzzleInput, ch)
		go ic.Run()
		<-ch
		ch <- i
		network[i] = ch
	}
	var packet, natPacket Packet
	var packets []Packet
	var ch chan int
	natHistory := make(map[int]struct{})

	for i := range 50 {
		ch = network[i]
		<-ch
		ch <- -1
		collectPackets(&packets, ch)
	}

	for {
		for len(packets) > 0 {
			packet = packets[0]
			packets = packets[1:]
			if packet.dst == 255 {
				natPacket = packet
				continue
			}
			ch = network[packet.dst]
			ch <- packet.x
			<-ch
			ch <- packet.y
			collectPackets(&packets, ch)
		}

		ch = network[0]
		if _, ok := natHistory[natPacket.y]; ok {
			return natPacket.y
		}
		natHistory[natPacket.y] = struct{}{}
		ch <- natPacket.x
		<-ch
		ch <- natPacket.y
		collectPackets(&packets, ch)
	}
}
